[Java Worker] Support raylet on Java

kingchin1218 opened the pull request:
<!--
Thank you for your contribution!

Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request.
-->

## What do these changes do?
Now only python worker supports raylet, and this change support raylet for java worker.
In lots of places, we use a flag `useRaylet` to distinguish raylet mode or not.

[How to enable?]
If you want to enable raylet on java, just make `use_raylet = true` in ray.config.ini file.
Other steps are the same as non-raylet mode.
 

Note:Our internal discussions are in this [PR](https://github.com/ant-tech-alliance/ray/pull/54). (:

<!-- Please give a short brief about these changes. -->

## Related issue number
N/A
<!-- Are there any issues opened that will be resolved by merging this change? -->

robertnishihara commented on src/ray/raylet/main.cc:
```diff
@@ -39,11 +40,8 @@ int main(int argc, char *argv[]) {
-    node_manager_config.worker_command.push_back(token);
-  }
+  boost::split(node_manager_config.worker_command, worker_command,
+               boost::is_any_of(" "), boost::token_compress_on);
```
Let's revert this. I think we want to minimize boost dependencies, so currently we just use `boost::asio`.

cc @pcmoritz

kingchin1218 replied:
@robertnishihara 
This small change have fixed a bug：
If we pass a command string like "a&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;b"(note:more than 1 space between a and b), this code will parse it into an array which contains `('a',' ', ' ',' ', ' ', 'b' )`. This will emit some errors when `startprocess(`('a',' ', ' ',' ', ' ', 'b' )`)` because of some useless spaces.
Therefore, we should compress some useless spaces when parse command string.

If we want to minimize boost dep, I will do this without it.

robertnishihara replied:
Oh that makes sense, good catch :) Still let's do this without boost.

robertnishihara commented on src/ray/raylet/main.cc:
```diff
@@ -83,8 +81,11 @@ int main(int argc, char *argv[]) {
+  auto handler = [&main_service, &raylet_socket_name](const boost::system::error_code &error,
+                                 int signal_number) {
+    main_service.stop();
+    remove(raylet_socket_name.c_str());
```
What does this line do?

kingchin1218 replied:
When raylet process exit safely, the local socket file like `raylet11111` is always kept there.

robertnishihara replied:
I see. Could you add a comment explaining that?

kingchin1218 replied:
My job.

robertnishihara commented on src/ray/raylet/task_spec.h:
```diff
@@ -108,6 +108,10 @@ class TaskSpecification {
                     int64_t num_returns,
                     const std::unordered_map<std::string, double> &required_resources);
+  
+  TaskSpecification(const uint8_t *spec, size_t spec_size) {
```
Document this constructor.

@stephanie-wang can you take a look at this?

robertnishihara commented on src/local_scheduler/local_scheduler_client.cc:
```diff
@@ -84,7 +84,7 @@ void local_scheduler_submit(LocalSchedulerConnection *conn,
     LocalSchedulerConnection *conn,
     const std::vector<ObjectID> &execution_dependencies,
-    ray::raylet::TaskSpecification task_spec) {
+    ray::raylet::TaskSpecification &task_spec) {
```
This can even by `const`, right?

kingchin1218 replied:
@robertnishihara  Yup.

robertnishihara commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -1,7 +1,6 @@
 #include "ray/raylet/worker_pool.h"
 
 #include <sys/wait.h>
-
```
I think we want this newline according to https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes.

kingchin1218 replied:
@robertnishihara  OK

robertnishihara commented on src/ray/raylet/main.cc:
```diff
@@ -83,8 +81,11 @@ int main(int argc, char *argv[]) {
+  auto handler = [&main_service, &raylet_socket_name](const boost::system::error_code &error,
+                                 int signal_number) {
+    main_service.stop();
+    remove(raylet_socket_name.c_str());
```
What does this line do?

kingchin1218 commented:
@robertnishihara Could you `@` some one to review the java code, or reiview it by urself?

robertnishihara commented:
@kingchin1218 you mean someone from Ant or outside of Ant?

raulchen commented:
I'll look at the Java part.

mylinyuzhi commented on java/runtime-native/src/main/java/org/ray/spi/impl/BaseStateStoreProxyImpl.java:
```diff
@@ -0,0 +1,112 @@
+        try {
+          RayLog.core.warn("Error occurred in BaseStateStoreProxyImpl getAddressInfo, "
+              + (numRetries - count) + " retries remaining", e);
+          TimeUnit.MILLISECONDS.sleep(1000);
```
it's better to wrap the call to Thread.sleep as an util func, avoid the InterruptedException everywhere，or like the guava's Uninterruptibles

kingchin1218 replied:
Good catch.

raulchen commented on java/runtime-native/src/main/java/org/ray/format/gcs/ClientTableData.java:
```diff
@@ -0,0 +1,79 @@
+package org.ray.format.gcs;
```
Don't commit this file.

kingchin1218 replied:
I saw that there are some auto-generated  files like this, like [arg.java](https://github.com/ray-project/ray/blob/master/java/runtime-native/src/main/java/org/ray/spi/impl/Arg.java).
I don't know whether we should commit this file in java worker, or make it built by the step  'build.sh'.

raulchen replied:
I think all auto-generated files shouldn't be committed. 
@salah-man is there any reason why Arg.java was committed? If not, let's remove it as well?

salah-man replied:
Yes, Arg.java could be generated totally, but the file of TaskInfo.java add function manually to avoid encoding and decoding cost, we could find it as the comment in TaskInfo.java.

raulchen commented on java/runtime-native/src/main/java/org/ray/runner/RunManager.java:
```diff
@@ -48,7 +48,7 @@ public RunManager(RayParameters params, PathConfig paths, ConfigReader configRea
   private static boolean killProcess(Process p) {
     if (p.isAlive()) {
-      p.destroyForcibly();
+      p.destroy();
```
Does this change fix any issue?

kingchin1218 replied:
Many backend processes are exited by a `SIGTERM` signal. When they do some tail-in works when receive the signal.
If we use `destroyForcibly()`, the backend processes will be killed without any signal. Therefore, the backend cannot be exited safely and do some tail-in works.

kingchin1218 commented:
@raulchen Thanks for ur help.

@robertnishihara : @raulchen will help reivew it.

raulchen commented on java/runtime-native/src/main/java/org/ray/spi/StateStoreProxy.java:
```diff
@@ -31,5 +31,19 @@
+   * @return A list of SchedulerInfo which contains raylet name and raylet port.
+   * @throws Exception No redis client exception.
+   */
+  List<AddressInfo> getAddressInfoHelper(final String nodeIpAddress,
```
This method shouldn't be part of the interface, because it's the implementation detail. I think it'd make more sense to make it a abstract method in `BaseStateStoreProxy`. (Maybe also give it a better name.)

kingchin1218 replied:
Good idea.

raulchen commented on java/runtime-native/src/main/java/org/ray/spi/impl/StateStoreProxyImpl.java:
```diff
@@ -1,201 +0,0 @@
-package org.ray.spi.impl;
```
It's better to do `git mv ../StateStoreProxyImpl.java ../BaseStateStoreProxyImpl.java`, instead of removing this file and creating a new file, so that git can still retain the history and github can compare these 2 files side by side.

kingchin1218 replied:
gotcha

raulchen commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -91,11 +90,13 @@ void WorkerPool::RegisterWorker(std::shared_ptr<Worker> worker) {
-  it->second--;
-  if (it->second == 0) {
-    starting_worker_processes_.erase(it);
+  if (it != starting_worker_processes_.end()) {
```
Why do we need this change?

kingchin1218 replied:
I have explained it in this [comment](https://github.com/ant-tech-alliance/ray/pull/54#discussion-diff-204663832L98).
@raulchen  cc

kingchin1218 commented:
I have addressed the comments.
@robertnishihara  @raulchen 
Could you please take a look at the whole PR again?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/6877/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/6885/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/6889/
Test PASSed.

robertnishihara commented:
I agree that we should not commit autogenerated files.

kingchin1218 commented:
@salah-man  @robertnishihara  @raulchen 
I found that there are at least 3 auto-generated files: `TaskInfo.java`, `ResourcePair.java`, `Arg.java`.
If we do this thing in this PR, I am afraid that this PR will be too heavy.
em, How about doing it in my next separate PR?

raulchen commented:
@kingchin1218 it's fine to remove those files in next PR

salah-man commented on java/runtime-common/src/main/java/org/ray/spi/LocalSchedulerProxy.java:
```diff
@@ -131,4 +135,33 @@ public void reconstructObject(UniqueID objectId) {
+    return ids;
+  }
+
+  public <T> WaitResult<T> wait(RayList<T> waitfor, int numReturns, int timeout) {
```
ObjectStoreProxy already have the wait functions, which from the JNI of plasma and open to the users by Ray.wait, so i am not sure the relationship between the wait function here in LocalSchedulerProxy and from JNI of ls, thanks.

salah-man replied:
And do not found any usage of this function.

salah-man replied:
Got it, it will be used for raylet.

kingchin1218 replied:
[This](https://github.com/ant-tech-alliance/ray/blob/java_raylet_2/java/runtime-common/src/main/java/org/ray/spi/ObjectStoreProxy.java#L103) invokes it.

salah-man commented:
LGTM!

robertnishihara commented:
@kingchin1218 why not remove `ClientTableData.java` from this PR? Will that cause any issues or errors? It's fine to remove the other autogenerated files in a separate PR, but I don't see a reason to add more autogenerated files in this PR. If doing so requires other code changes then we can do that in a separate PR.

Also, would you mind rebasing this PR? It looks like I introduced a small conflict when merging #2362.

kingchin1218 commented:
@robertnishihara 
We have discussed the auto-gen files issue internally:
We should add some code in CMAKELISTS.txt to build the auto-files before we remove it, otherwise the building will failed.
Therefore, I added the code to build the auto-gen files and removed all of the auto-gen files in my local.
Once this is merged, I will request another PR to remove them.

robertnishihara commented:
@kingchin1218 ok, that sounds good.

kingchin1218 commented:
I will reslove the conflicts right now.

robertnishihara commented:
@kingchin1218 just a reminder about this :) Let me know if you run into any issues.

kingchin1218 commented:
@robertnishihara 
I'm sorry that there is an issue while i am merge it into [Stephanie's PR](https://github.com/ray-project/ray/pull/2362).

After resloved the conflicts, I ran the tests with `mvn test` command. Unfortunately, when it executed [`ActorTest.java`](https://github.com/ray-project/ray/blob/master/java/test/src/main/java/org/ray/api/test/ActorTest.java), node manager crashed at [HandleObjectLocal line 1010](https://github.com/ray-project/ray/pull/2362/files#diff-50e88810cd198c83bbce3317b1bc9924R1010). 

I found that when node manager calls `HandleObjectLocal()` some time,The  tasks(named `TASKA`) in `ready_task_id_set` are actually removed from `waiting queue` .
It does make sense now.
However, when the func `HandleObjectLocal()` is called again, `ready_task_id_set` also contains 'TASKA'.
but now, `TASKA` is in `running queue`. As a result, `TASKA` is still in `ready_task_id_set` , and `RAY_CHECK` fails.

I think I need @stephanie-wang 's help.  Could u(@stephanie-wang ) please tell me which reasons may make this issue?

```c++
void NodeManager::HandleObjectLocal(const ObjectID &object_id) {
  // Notify the task dependency manager that this object is local.
  const auto ready_task_ids = task_dependency_manager_.HandleObjectLocal(object_id);
  // Transition the tasks whose dependencies are now fulfilled to the ready
  // state.
  if (ready_task_ids.size() > 0) {
    std::unordered_set<TaskID> ready_task_id_set(ready_task_ids.begin(),
                                                 ready_task_ids.end());
    // Transition tasks from waiting to scheduled.
    local_queues_.MoveTasks(ready_task_id_set, TaskState::WAITING, TaskState::READY);
    // New scheduled tasks appeared in the queue, try to dispatch them.
    DispatchTasks();

    // Check that remaining tasks that could not be transitioned are blocked
    // workers or drivers.
    local_queues_.FilterState(ready_task_id_set, TaskState::BLOCKED);
    local_queues_.FilterState(ready_task_id_set, TaskState::DRIVER);
    RAY_CHECK(ready_task_id_set.empty());
  }
}
```

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7024/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7022/
Test PASSed.

raulchen commented on java/runtime-common/src/main/java/org/ray/core/RayRuntime.java:
```diff
@@ -374,16 +374,16 @@ public RayObjects call(UniqueID taskId, Class<?> funcCls, RayFunc lambda, int re
             isMetadata);
       }
-      RayLog.core.debug(
+      RayLog.core.warn(
```
why change this?

kingchin1218 replied:
I have not changed it. It should be master branch changed it into `warn`.
I'am going to sync this to master.

raulchen commented on java/runtime-common/src/main/java/org/ray/spi/LocalSchedulerLink.java:
```diff
@@ -15,19 +15,19 @@
 
-  default void fetch(UniqueID objectId) {
-    reconstructObject(objectId);
+  default void fetch(UniqueID objectId, boolean fetchOnly) {
```
This method shouldn't have the `fetchOnly` parameter.

In Raylet's local scheduler client, `fetch = reconstruct_object(fetch_only=True)`. When `fetch_only=False`, `reconstruct` = `reconstruct and fetch`.

I found the meaning of the this function is confusing, maybe we should separate them to 2 functions `fetch` and `reconstruct_and_fetch`? @robertnishihara

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7039/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7047/
Test PASSed.

kingchin1218 commented:
@robertnishihara  @raulchen 
CC :)

raulchen commented on java/runtime-common/src/main/java/org/ray/core/RayRuntime.java:
```diff
@@ -357,69 +349,53 @@ public RayObjects call(UniqueID taskId, Class<?> funcCls, RayFunc lambda, int re
-    }
-
+    assert results.size() == 1;
+    return results.get(0);
```
minor, this can be simplified to one line `return doGet(ImmutableList(objectId, isMetadata).get(0)`.
but if you prefer the current way, it's fine as well.

raulchen commented on java/runtime-common/src/main/java/org/ray/core/RayRuntime.java:
```diff
@@ -357,69 +349,53 @@ public RayObjects call(UniqueID taskId, Class<?> funcCls, RayFunc lambda, int re
   // We divide the fetch into smaller fetches so as to not block the manager
   // for a prolonged period of time in a single call.
   private void dividedFetch(List<UniqueID> objectIds) {
-    int fetchSize = objectStoreProxy.getFetchSize();
```
is `getFetchSize()` still used by any place? if not, let's remove it.

raulchen commented on java/runtime-common/src/main/java/org/ray/core/RayRuntime.java:
```diff
@@ -357,69 +349,53 @@ public RayObjects call(UniqueID taskId, Class<?> funcCls, RayFunc lambda, int re
+
+      List<UniqueID> fetchIds = (endIndex < numObjectIds)
+          ? objectIds.subList(i, endIndex)
+          : objectIds.subList(i, numObjectIds);
```
nit, I think moving the question mark to the end of the above line would improve readability.

raulchen commented on java/runtime-common/src/main/java/org/ray/core/RayRuntime.java:
```diff
@@ -357,69 +349,53 @@ public RayObjects call(UniqueID taskId, Class<?> funcCls, RayFunc lambda, int re
+    }
+  }
+
+  private void dividedReconstruct(List<UniqueID> objectIds) {
```
better unify this method with `dividedFetch` and remove duplicate code.
One possible way is using a helper function that splits the list into small batches (returning `List<List<UniqueID>>`, and still call `reconstructObjects/fetch` in `doGet`

raulchen commented on java/runtime-common/src/main/java/org/ray/core/RayRuntime.java:
```diff
@@ -357,69 +349,53 @@ public RayObjects call(UniqueID taskId, Class<?> funcCls, RayFunc lambda, int re
+          ? objectIds.subList(i, endIndex)
+          : objectIds.subList(i, numObjectIds);
+
+      localSchedulerProxy.reconstructObjects(reconstructIds, false);
```
In legacy code, `reconstructObjects` can only handle one ID at a time, we should use a loop here.

raulchen commented on java/runtime-common/src/main/java/org/ray/spi/LocalSchedulerLink.java:
```diff
@@ -17,17 +17,9 @@
 
   void reconstructObject(UniqueID objectId, boolean fetchOnly);
 
+  void reconstructObjects(byte[][] objectIds, boolean fetchOnly);
```
we should make the parameter type of these 2 method consistent, use `List<UniqueID>`.

raulchen commented on java/runtime-native/src/main/java/org/ray/spi/impl/DefaultLocalSchedulerClient.java:
```diff
@@ -92,14 +92,31 @@ public void reconstructObject(UniqueID objectId, boolean fetchOnly) {
+      builder.append(" ");
+    }
+
+    RayLog.core.info("reconstruct objects " + builder.toString());
```
We should avoid building the string here, because it's unnecessary when info-level logging isn't enabled.
If you change the parameter type to `List<UnqiueID>`, then you can just do `RayLog.core.info("Reconstructing objects {}", objectIds);` 
（Note: here we use `{}` as a placeholder to avoid unnecessary string concatenation. It's a legacy issue that most of the code in Ray isn't using this patten. If you don't mind, could you also fix other places related in this PR?)

raulchen commented on src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc:
```diff
@@ -208,6 +208,33 @@ Java_org_ray_spi_impl_DefaultLocalSchedulerClient__1reconstruct_1object(
+ * Signature: (J[B)V
+ */
+JNIEXPORT void JNICALL
+Java_org_ray_spi_impl_DefaultLocalSchedulerClient__1reconstruct_1objects(
```
Let's only keep this plural version in c++ code, and have the java single version call the java plural version.

raulchen commented:
@kingchin1218, thanks! this PR is pretty good right now, I just left a few small comments, could you fix them? then I think it'd be good for merge.

raulchen commented:
hi @robertnishihara, a quick question. This piece of code isn't needed for raylet, right? 
https://github.com/ray-project/ray/blob/d9a019b8e5a722c3aef7ae9094ad53dce92f6410/python/ray/worker.py#L2122-L2156

kingchin1218 commented:
@raulchen  @robertnishihara 
Could you please cc? :)

kingchin1218 commented:
@raulchen  Thanks for helping me to review it.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7058/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7059/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7060/
Test FAILed.

kingchin1218 commented:
Jenkins,  retest.

kingchin1218 commented:
jenkins, retest please. :)

robertnishihara commented on src/ray/raylet/task_spec.h:
```diff
@@ -109,6 +109,14 @@ class TaskSpecification {
+  ///
+  /// \param spec The specification data from a pointer.
+  /// \param spec_size The specification data size.
+  TaskSpecification(const uint8_t *spec, size_t spec_size) {
```
Actually, we're doing the same thing in #2526 as well. We should make this

```cpp
TaskSpecification(const std::string &string);
```

kingchin1218 replied:
I think it is not conflict.
If we only use `TaskSpecification(const std::string &string);`
we must pack the pointer into a string [here](https://github.com/ant-tech-alliance/ray/blob/java_raylet_2/src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc#L93), and call  string-verson ctor. 
but in string-version ctor, we should unpack it to data-pointer.
I think we should remaining the 2 version ctor, string-version and data-pointer version.
How do you like it? Or do u have a better idea? @robertnishihara

kingchin1218 replied:
It' ok. :)

kingchin1218 commented:
@robertnishihara  It have already been addressed, Could u take a look ?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7088/
Test FAILed.

robertnishihara commented:
@kingchin1218 thanks a lot! Sorry about the delay.
